### PR TITLE
Cross-dataset: torch.compile throughput — 20-40% speedup, more epochs per wall-clock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

`torch.compile` (PyTorch 2.x) fuses operations and eliminates Python overhead, typically giving 20-40% throughput improvement on transformer models with no change to model quality. The `--compile-model` flag already exists in `train.py` but has not been tested at scale across all three datasets. With our fixed epoch budget, more throughput = more gradient steps per wall-clock hour = better-converged models within the same SENPAI timeout. This is a free win if it works.

## Instructions

The `--compile-model` flag already exists. Just add it to the golden configs for all three datasets and run 6 experiments (2 per dataset: one with compile, one without, to measure throughput delta).

Use `--wandb_group compile-throughput` for all runs.

**DrivAerML — with compile:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --compile-model --wandb_group compile-throughput
```

**DrivAerML — without compile (control):**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb_group compile-throughput
```

**AirfRANS — with compile:**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --compile-model --wandb_group compile-throughput
```

**AirfRANS — without compile (control):**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb_group compile-throughput
```

**TandemFoil — with compile:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --compile-model --wandb_group compile-throughput
```

**TandemFoil — without compile (control):**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb_group compile-throughput
```

In your PR comment, please report:
- Steps/sec (or epochs/hour) with and without compile for each dataset
- Final val/loss for each run
- Any compile errors or warnings encountered

## Baseline

Current best metrics:
- **TandemFoil**: val/loss = 30.10 (epoch 157, PR #2810, W&B run v6amjkh7)
- **AirfRANS**: val/loss = 0.001236 (epoch 349, PR #2828, W&B run libpwryz)
- **DrivAerML**: val/loss = 4.619% (epoch 256, PR #2691, W&B run k8qtsxxz)

Reproduce commands:
```bash
# TandemFoil baseline
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999

# AirfRANS baseline
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999

# DrivAerML baseline
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT piccolo:

## Results

### Throughput Summary

| Dataset | Compile | No-Compile | Throughput Gain | Epochs (same wall time) |
|---------|---------|------------|-----------------|-------------------------|
| **AirfRANS** | 28.6 sec/epoch | 42.3 sec/epoch | **+47.7%** | 755 vs 511 |
| **TandemFoil** | 66.3 sec/epoch | 90.6 sec/epoch | **+36.7%** | 328 vs 240 |
| **DrivAerML** | 44.9 sec/epoch | 44.9 sec/epoch | **0%** | 482 vs 482 |

### Val/Loss Summary

| Dataset | Metric | Compile (best) | No-Compile (best) | Baseline |
|---------|--------|----------------|---------------------|----------|
| **AirfRANS** | val_primary/surface_mse | **0.000999** (ep 489) | 0.001051 (ep 365) | 0.001236 |
| **TandemFoil** | val_primary/surface_pressure_mae | **27.50** (ep 291) | 27.72 (ep 229) | 30.10 |
| **DrivAerML** | val_primary/surface_rel_l2_pct | 12.66% (ep 264) ⚠️ NaN@454 | **3.997%** (ep 467) | 4.619% |

### Test Metrics

| Dataset | Metric | Compile (test) | No-Compile (test) |
|---------|--------|----------------|---------------------|
| **AirfRANS** | test_primary/surface_mse | **0.002415** | 0.005022 |
| **TandemFoil** | test_primary/surface_pressure_mae | **32.13** | 39.34 |
| **DrivAerML** | test_primary/surface_rel_l2_pct | NaN ⚠️ | **5.932%** |

### DrivAerML Compile Divergence

The DrivAerML compile run diverged to NaN at epoch 454. Gradient norms exploded (14T → 25T → NaN) before all metrics went NaN. The no-compile control run trained stably through all 482 epochs. Note: DrivAerML config does not use `--grad-clip`, which may make it more susceptible to precision differences from operator fusion. Also noteworthy: the default `compile_model: bool = True` means the existing baseline (4.619%) was technically run with compile — this divergence may be stochastic or environment-dependent.

Despite no throughput gain (same epochs/time), the compile run also had numerically different training dynamics — worse val/loss throughout (never reached below 12.66%) compared to no-compile (3.997%).

### Key Findings

1. **AirfRANS: Strong win.** 48% more epochs in the same wall time, translating to 19% better val vs baseline (0.000999 vs 0.001236) and 52% better test metric.
2. **TandemFoil: Strong win.** 37% more epochs, 8.6% better val vs baseline (27.50 vs 30.10), 18% better test.
3. **DrivAerML: Compile is harmful.** Zero throughput gain and caused numerical instability (NaN divergence). The no-compile run actually beat baseline (3.997% vs 4.619%).

### No compile errors or warnings

No compile-time errors or warnings were encountered. All `torch.compile` invocations succeeded. The DrivAerML issue is a runtime numerical stability problem, not a compile error.

### Peak Memory

The codebase does not currently log peak GPU memory. All runs fit comfortably in the 98GB RTX PRO 6000 VRAM.

### Commands Used

```bash
# DrivAerML — with compile (GPU 0)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --compile-model --agent piccolo --wandb-name "piccolo/drivaerml-compile" --wandb-group compile-throughput

# DrivAerML — without compile (GPU 1)
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --no-compile-model --agent piccolo --wandb-name "piccolo/drivaerml-no-compile" --wandb-group compile-throughput

# AirfRANS — with compile (GPU 2)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --compile-model --agent piccolo --wandb-name "piccolo/airfrans-compile" --wandb-group compile-throughput

# AirfRANS — without compile (GPU 3)
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --no-compile-model --agent piccolo --wandb-name "piccolo/airfrans-no-compile" --wandb-group compile-throughput

# TandemFoil — with compile (GPU 4)
CUDA_VISIBLE_DEVICES=4 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --compile-model --agent piccolo --wandb-name "piccolo/tandemfoil-compile" --wandb-group compile-throughput

# TandemFoil — without compile (GPU 5)
CUDA_VISIBLE_DEVICES=5 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --no-compile-model --agent piccolo --wandb-name "piccolo/tandemfoil-no-compile" --wandb-group compile-throughput
```

### W&B Runs

| Run | W&B ID |
|-----|--------|
| DrivAerML compile | `0orlg9gc` |
| DrivAerML no-compile | `bht6h42t` |
| AirfRANS compile | `p1n6vh3u` |
| AirfRANS no-compile | `lcli8hhy` |
| TandemFoil compile | `0gu790cw` |
| TandemFoil no-compile | `bblxmkpd` |

### What happened

`torch.compile` delivers substantial throughput gains for AirfRANS (+48%) and TandemFoil (+37%), which directly translate to better metrics because the model gets more gradient steps within the same wall-clock budget. However, DrivAerML saw zero throughput benefit and compile caused catastrophic NaN divergence, likely due to precision differences from operator fusion in the absence of gradient clipping.

**Important note:** The PR's control commands (without compile) did not explicitly disable compile, but `compile_model` defaults to `True` in the config. I added `--no-compile-model` to the control runs to ensure a true comparison.

### Suggested follow-ups

1. **Enable `--grad-clip` for DrivAerML with compile** — the divergence likely stems from unclipped gradient spikes amplified by compile's fused operations. Adding `--grad-clip 1.0` (as used by AirfRANS and TandemFoil) could stabilize the DrivAerML compile run.
2. **Make compile the default for AirfRANS and TandemFoil golden configs** — the throughput gain is substantial and consistent.
3. **Investigate DrivAerML compile throughput** — it's surprising that DrivAerML sees zero speedup while the other datasets get 37-48%. This could be related to the batch_size=1 + 50k surface points pattern creating different operator shapes that don't benefit from compile.
4. **Add peak GPU memory logging** to train.py for future experiments (`torch.cuda.max_memory_allocated()`).

**Comment by tcapelle** [comment]
ADVISOR: Merging this PR — the DrivAerML no-compile run achieved **3.997%** val surface_rel_l2_pct (ep467, W&B: bht6h42t), a genuine -13.5% improvement over the 4.619% baseline. The key finding is that torch.compile provides no throughput benefit on DrivAerML but does not hurt either, while the baseline config with proper training duration beats the old best. The compile run unfortunately diverged (NaN at ep454) due to missing --grad-clip — future DrivAerML compile experiments must include --grad-clip 1.0. Merging the no-compile result as the new DrivAerML champion.
